## import libraries

In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv('/content/drive/My Drive/machine learning/classification/data/Wine_Quality_Data.csv')

In [3]:
df.head()

fixed_acidity  volatile_acidity  citric_acid  ...  alcohol  quality  color
0            7.4              0.70         0.00  ...      9.4        5    red
1            7.8              0.88         0.00  ...      9.8        5    red
2            7.8              0.76         0.04  ...      9.8        5    red
3           11.2              0.28         0.56  ...      9.8        6    red
4            7.4              0.70         0.00  ...      9.4        5    red

[5 rows x 13 columns]

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6497 entries, 0 to 6496
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed_acidity         6497 non-null   float64
 1   volatile_acidity      6497 non-null   float64
 2   citric_acid           6497 non-null   float64
 3   residual_sugar        6497 non-null   float64
 4   chlorides             6497 non-null   float64
 5   free_sulfur_dioxide   6497 non-null   float64
 6   total_sulfur_dioxide  6497 non-null   float64
 7   density               6497 non-null   float64
 8   pH                    6497 non-null   float64
 9   sulphates             6497 non-null   float64
 10  alcohol               6497 non-null   float64
 11  quality               6497 non-null   int64  
 12  color                 6497 non-null   object 
dtypes: float64(11), int64(1), object(1)
memory usage: 660.0+ KB


In [5]:
## we have 1 object column

# we have to check how many categorical values are present

In [6]:
df.color.unique()

array(['red', 'white'], dtype=object)

In [7]:
# in this the only target column is in categorical column
# we are going to change the column into numerical by using LabelEncoder

In [8]:
from sklearn.preprocessing import  LabelEncoder
encoder = LabelEncoder()

In [9]:
df.iloc[:,12]=encoder.fit_transform(df.iloc[:,12])

In [10]:
df.head()

fixed_acidity  volatile_acidity  citric_acid  ...  alcohol  quality  color
0            7.4              0.70         0.00  ...      9.4        5      0
1            7.8              0.88         0.00  ...      9.8        5      0
2            7.8              0.76         0.04  ...      9.8        5      0
3           11.2              0.28         0.56  ...      9.8        6      0
4            7.4              0.70         0.00  ...      9.4        5      0

[5 rows x 13 columns]

In [11]:
encoder.classes_

array(['red', 'white'], dtype=object)

In [12]:
df.color.unique()

array([0, 1])

In [13]:
df.color.value_counts()

1    4898
0    1599
Name: color, dtype: int64

In [14]:
# create feature and target labels

In [15]:
X= df.iloc[:,:12].values
y = df.iloc[:,12].values

In [16]:
X

array([[ 7.4 ,  0.7 ,  0.  , ...,  0.56,  9.4 ,  5.  ],
       [ 7.8 ,  0.88,  0.  , ...,  0.68,  9.8 ,  5.  ],
       [ 7.8 ,  0.76,  0.04, ...,  0.65,  9.8 ,  5.  ],
       ...,
       [ 6.5 ,  0.24,  0.19, ...,  0.46,  9.4 ,  6.  ],
       [ 5.5 ,  0.29,  0.3 , ...,  0.38, 12.8 ,  7.  ],
       [ 6.  ,  0.21,  0.38, ...,  0.32, 11.8 ,  6.  ]])

In [17]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size = 0.2,random_state = 42)

In [18]:
from sklearn.linear_model import LogisticRegression

In [19]:
lr = LogisticRegression()
lr.fit(X_train,y_train)
lr.score(X_test,y_test)

/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


0.9730769230769231

In [20]:
from sklearn.preprocessing import MinMaxScaler
scale = MinMaxScaler()

In [21]:
X_strain = scale.fit_transform(X_train)
X_stest = scale.transform(X_test)

In [22]:
lr.fit(X_strain,y_train)
lr.score(X_stest,y_test)

0.9830769230769231

In [23]:
lr.get_params

<bound method BaseEstimator.get_params of LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)>

In [24]:
# WE CHECK THE CONFUSION MATRIX

In [25]:
from sklearn.metrics import classification_report,confusion_matrix

In [26]:
y_pred = lr.predict(X_stest)
print(confusion_matrix(y_test,y_pred))

[[323  18]
 [  4 955]]


In [27]:
classification_report(y_test,y_pred)

'              precision    recall  f1-score   support\n\n           0       0.99      0.95      0.97       341\n           1       0.98      1.00      0.99       959\n\n    accuracy                           0.98      1300\n   macro avg       0.98      0.97      0.98      1300\nweighted avg       0.98      0.98      0.98      1300\n'

              precision    recall  f1-score   support

           0       0.99      0.95      0.97       341
           1       0.98      1.00      0.99       959

     accuracy                          0.98      1300
    macro avg      0.98      0.97      0.98      1300
    weighted avg   0.98      0.98      0.98      1300


In [28]:
# this data come without hyperparameter tuning

In [30]:
lr= LogisticRegression(penalty='l2',C=10)
lr.fit(X_strain,y_train)
lr.score(X_stest,y_test)

0.9853846153846154

In [31]:
from sklearn.model_selection import GridSearchCV

In [32]:
penalty = ['l1','l2']
C = np.logspace(-4,4,20)
solver = ['liblinear','sag','lbfgs']

parameter ={
    'penalty':penalty,
    'C':C,
    'solver':solver
}
lr = LogisticRegression(random_state=42)

In [33]:
grid_search = GridSearchCV(lr,param_grid=parameter,cv=3,verbose=3,)

In [34]:
grid_search.fit(X_strain,y_train)

Fitting 3 folds for each of 120 candidates, totalling 360 fits
[CV] C=0.0001, penalty=l1, solver=liblinear ..........................
[CV]  C=0.0001, penalty=l1, solver=liblinear, score=0.242, total=   0.0s
[CV] C=0.0001, penalty=l1, solver=liblinear ..........................
[CV]  C=0.0001, penalty=l1, solver=liblinear, score=0.242, total=   0.0s
[CV] C=0.0001, penalty=l1, solver=liblinear ..........................
[CV]  C=0.0001, penalty=l1, solver=liblinear, score=0.242, total=   0.0s
[CV] C=0.0001, penalty=l1, solver=sag ................................
[CV] ...... C=0.0001, penalty=l1, solver=sag, score=nan, total=   0.0s
[CV] C=0.0001, penalty=l1, solver=sag ................................
[CV] ...... C=0.0001, penalty=l1, solver=sag, score=nan, total=   0.0s
[CV] C=0.0001, penalty=l1, solver=sag ................................
[CV] ...... C=0.0001, penalty=l1, solver=sag, score=nan, total=   0.0s
[CV] C=0.0001, penalty=l1, solver=lbfgs ..............................
[CV] ...

[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed:    0.0s remaining:    0.0s
/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_validation.py:536: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
ValueError: Solver sag supports only 'l2' or 'none' penalties, got l1 penalty.

  FitFailedWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_validation.py:536: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
ValueError: Solver lbfgs supports only 'l2' or 'none' penalties, got l1 penalty.

  FitFailedWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_validation.py:536: FitFailedWarning: Estimator fit failed. The scor

[CV]  C=0.00026366508987303583, penalty=l2, solver=liblinear, score=0.758, total=   0.0s
[CV] C=0.00026366508987303583, penalty=l2, solver=sag ................
[CV]  C=0.00026366508987303583, penalty=l2, solver=sag, score=0.758, total=   0.0s
[CV] C=0.00026366508987303583, penalty=l2, solver=sag ................
[CV]  C=0.00026366508987303583, penalty=l2, solver=sag, score=0.758, total=   0.0s
[CV] C=0.00026366508987303583, penalty=l2, solver=sag ................
[CV]  C=0.00026366508987303583, penalty=l2, solver=sag, score=0.758, total=   0.0s
[CV] C=0.00026366508987303583, penalty=l2, solver=lbfgs ..............
[CV]  C=0.00026366508987303583, penalty=l2, solver=lbfgs, score=0.758, total=   0.0s
[CV] C=0.00026366508987303583, penalty=l2, solver=lbfgs ..............
[CV]  C=0.00026366508987303583, penalty=l2, solver=lbfgs, score=0.758, total=   0.0s
[CV] C=0.00026366508987303583, penalty=l2, solver=lbfgs ..............
[CV]  C=0.00026366508987303583, penalty=l2, solver=lbfgs, score=0.

/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_validation.py:536: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
ValueError: Solver sag supports only 'l2' or 'none' penalties, got l1 penalty.

  FitFailedWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_validation.py:536: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
ValueError: Solver lbfgs supports only 'l2' or 'none' penalties, got l1 penalty.

  FitFailedWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_validation.py:536: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
ValueError: Solver sag supports only 'l2' or 'none' penalties, got l1 penalty.

  FitFailedWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/model_select

[CV]  C=0.0006951927961775605, penalty=l2, solver=sag, score=0.758, total=   0.0s
[CV] C=0.0006951927961775605, penalty=l2, solver=lbfgs ...............
[CV]  C=0.0006951927961775605, penalty=l2, solver=lbfgs, score=0.758, total=   0.0s
[CV] C=0.0006951927961775605, penalty=l2, solver=lbfgs ...............
[CV]  C=0.0006951927961775605, penalty=l2, solver=lbfgs, score=0.758, total=   0.0s
[CV] C=0.0006951927961775605, penalty=l2, solver=lbfgs ...............
[CV]  C=0.0006951927961775605, penalty=l2, solver=lbfgs, score=0.758, total=   0.0s
[CV] C=0.0018329807108324356, penalty=l1, solver=liblinear ...........
[CV]  C=0.0018329807108324356, penalty=l1, solver=liblinear, score=0.758, total=   0.0s
[CV] C=0.0018329807108324356, penalty=l1, solver=liblinear ...........
[CV]  C=0.0018329807108324356, penalty=l1, solver=liblinear, score=0.758, total=   0.0s
[CV] C=0.0018329807108324356, penalty=l1, solver=liblinear ...........
[CV]  C=0.0018329807108324356, penalty=l1, solver=liblinear, sco

/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_validation.py:536: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
ValueError: Solver sag supports only 'l2' or 'none' penalties, got l1 penalty.

  FitFailedWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_validation.py:536: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
ValueError: Solver lbfgs supports only 'l2' or 'none' penalties, got l1 penalty.

  FitFailedWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_validation.py:536: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
ValueError: Solver sag supports only 'l2' or 'none' penalties, got l1 penalty.

  FitFailedWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/model_select

[CV]  C=0.012742749857031334, penalty=l2, solver=liblinear, score=0.774, total=   0.0s
[CV] C=0.012742749857031334, penalty=l2, solver=liblinear ............
[CV]  C=0.012742749857031334, penalty=l2, solver=liblinear, score=0.770, total=   0.0s
[CV] C=0.012742749857031334, penalty=l2, solver=liblinear ............
[CV]  C=0.012742749857031334, penalty=l2, solver=liblinear, score=0.779, total=   0.0s
[CV] C=0.012742749857031334, penalty=l2, solver=sag ..................
[CV]  C=0.012742749857031334, penalty=l2, solver=sag, score=0.766, total=   0.0s
[CV] C=0.012742749857031334, penalty=l2, solver=sag ..................
[CV]  C=0.012742749857031334, penalty=l2, solver=sag, score=0.761, total=   0.0s
[CV] C=0.012742749857031334, penalty=l2, solver=sag ..................
[CV]  C=0.012742749857031334, penalty=l2, solver=sag, score=0.767, total=   0.0s
[CV] C=0.012742749857031334, penalty=l2, solver=lbfgs ................
[CV]  C=0.012742749857031334, penalty=l2, solver=lbfgs, score=0.766, t

/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_validation.py:536: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
ValueError: Solver sag supports only 'l2' or 'none' penalties, got l1 penalty.

  FitFailedWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_validation.py:536: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
ValueError: Solver lbfgs supports only 'l2' or 'none' penalties, got l1 penalty.

  FitFailedWarning)



[CV] C=0.03359818286283781, penalty=l2, solver=sag ...................
[CV]  C=0.03359818286283781, penalty=l2, solver=sag, score=0.900, total=   0.0s
[CV] C=0.03359818286283781, penalty=l2, solver=sag ...................
[CV]  C=0.03359818286283781, penalty=l2, solver=sag, score=0.905, total=   0.0s
[CV] C=0.03359818286283781, penalty=l2, solver=sag ...................
[CV]  C=0.03359818286283781, penalty=l2, solver=sag, score=0.907, total=   0.0s
[CV] C=0.03359818286283781, penalty=l2, solver=lbfgs .................
[CV]  C=0.03359818286283781, penalty=l2, solver=lbfgs, score=0.900, total=   0.0s
[CV] C=0.03359818286283781, penalty=l2, solver=lbfgs .................
[CV]  C=0.03359818286283781, penalty=l2, solver=lbfgs, score=0.905, total=   0.0s
[CV] C=0.03359818286283781, penalty=l2, solver=lbfgs .................
[CV]  C=0.03359818286283781, penalty=l2, solver=lbfgs, score=0.907, total=   0.0s
[CV] C=0.08858667904100823, penalty=l1, solver=liblinear .............
[CV]  C=0.088586

/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_validation.py:536: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
ValueError: Solver sag supports only 'l2' or 'none' penalties, got l1 penalty.

  FitFailedWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_validation.py:536: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
ValueError: Solver lbfgs supports only 'l2' or 'none' penalties, got l1 penalty.

  FitFailedWarning)


[CV]  C=0.08858667904100823, penalty=l2, solver=liblinear, score=0.963, total=   0.0s
[CV] C=0.08858667904100823, penalty=l2, solver=liblinear .............
[CV]  C=0.08858667904100823, penalty=l2, solver=liblinear, score=0.969, total=   0.0s
[CV] C=0.08858667904100823, penalty=l2, solver=liblinear .............
[CV]  C=0.08858667904100823, penalty=l2, solver=liblinear, score=0.958, total=   0.0s
[CV] C=0.08858667904100823, penalty=l2, solver=sag ...................
[CV]  C=0.08858667904100823, penalty=l2, solver=sag, score=0.960, total=   0.0s
[CV] C=0.08858667904100823, penalty=l2, solver=sag ...................
[CV]  C=0.08858667904100823, penalty=l2, solver=sag, score=0.971, total=   0.0s
[CV] C=0.08858667904100823, penalty=l2, solver=sag ...................
[CV]  C=0.08858667904100823, penalty=l2, solver=sag, score=0.959, total=   0.0s
[CV] C=0.08858667904100823, penalty=l2, solver=lbfgs .................
[CV]  C=0.08858667904100823, penalty=l2, solver=lbfgs, score=0.960, total=  

/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_validation.py:536: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
ValueError: Solver sag supports only 'l2' or 'none' penalties, got l1 penalty.

  FitFailedWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_validation.py:536: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
ValueError: Solver lbfgs supports only 'l2' or 'none' penalties, got l1 penalty.

  FitFailedWarning)


[CV]  C=0.23357214690901212, penalty=l2, solver=sag, score=0.974, total=   0.0s
[CV] C=0.23357214690901212, penalty=l2, solver=lbfgs .................
[CV]  C=0.23357214690901212, penalty=l2, solver=lbfgs, score=0.977, total=   0.0s
[CV] C=0.23357214690901212, penalty=l2, solver=lbfgs .................
[CV]  C=0.23357214690901212, penalty=l2, solver=lbfgs, score=0.986, total=   0.0s
[CV] C=0.23357214690901212, penalty=l2, solver=lbfgs .................
[CV]  C=0.23357214690901212, penalty=l2, solver=lbfgs, score=0.974, total=   0.0s
[CV] C=0.615848211066026, penalty=l1, solver=liblinear ...............
[CV]  C=0.615848211066026, penalty=l1, solver=liblinear, score=0.987, total=   0.1s
[CV] C=0.615848211066026, penalty=l1, solver=liblinear ...............
[CV]  C=0.615848211066026, penalty=l1, solver=liblinear, score=0.991, total=   0.1s
[CV] C=0.615848211066026, penalty=l1, solver=liblinear ...............
[CV]  C=0.615848211066026, penalty=l1, solver=liblinear, score=0.986, total=   0

/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_validation.py:536: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
ValueError: Solver sag supports only 'l2' or 'none' penalties, got l1 penalty.

  FitFailedWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_validation.py:536: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
ValueError: Solver lbfgs supports only 'l2' or 'none' penalties, got l1 penalty.

  FitFailedWarning)


[CV]  C=0.615848211066026, penalty=l2, solver=lbfgs, score=0.986, total=   0.0s
[CV] C=0.615848211066026, penalty=l2, solver=lbfgs ...................
[CV]  C=0.615848211066026, penalty=l2, solver=lbfgs, score=0.990, total=   0.0s
[CV] C=0.615848211066026, penalty=l2, solver=lbfgs ...................
[CV]  C=0.615848211066026, penalty=l2, solver=lbfgs, score=0.979, total=   0.0s
[CV] C=1.623776739188721, penalty=l1, solver=liblinear ...............
[CV]  C=1.623776739188721, penalty=l1, solver=liblinear, score=0.990, total=   0.1s
[CV] C=1.623776739188721, penalty=l1, solver=liblinear ...............
[CV]  C=1.623776739188721, penalty=l1, solver=liblinear, score=0.994, total=   0.1s
[CV] C=1.623776739188721, penalty=l1, solver=liblinear ...............
[CV]  C=1.623776739188721, penalty=l1, solver=liblinear, score=0.991, total=   0.1s
[CV] C=1.623776739188721, penalty=l1, solver=sag .....................
[CV]  C=1.623776739188721, penalty=l1, solver=sag, score=nan, total=   0.0s
[CV] C

/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_validation.py:536: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
ValueError: Solver sag supports only 'l2' or 'none' penalties, got l1 penalty.

  FitFailedWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_validation.py:536: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
ValueError: Solver lbfgs supports only 'l2' or 'none' penalties, got l1 penalty.

  FitFailedWarning)


[CV]  C=1.623776739188721, penalty=l2, solver=sag, score=0.983, total=   0.0s
[CV] C=1.623776739188721, penalty=l2, solver=lbfgs ...................
[CV]  C=1.623776739188721, penalty=l2, solver=lbfgs, score=0.987, total=   0.0s
[CV] C=1.623776739188721, penalty=l2, solver=lbfgs ...................
[CV]  C=1.623776739188721, penalty=l2, solver=lbfgs, score=0.991, total=   0.0s
[CV] C=1.623776739188721, penalty=l2, solver=lbfgs ...................
[CV]  C=1.623776739188721, penalty=l2, solver=lbfgs, score=0.983, total=   0.0s
[CV] C=4.281332398719396, penalty=l1, solver=liblinear ...............
[CV]  C=4.281332398719396, penalty=l1, solver=liblinear, score=0.992, total=   0.1s
[CV] C=4.281332398719396, penalty=l1, solver=liblinear ...............
[CV]  C=4.281332398719396, penalty=l1, solver=liblinear, score=0.995, total=   0.1s
[CV] C=4.281332398719396, penalty=l1, solver=liblinear ...............
[CV]  C=4.281332398719396, penalty=l1, solver=liblinear, score=0.994, total=   0.1s
[CV]

/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_validation.py:536: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
ValueError: Solver sag supports only 'l2' or 'none' penalties, got l1 penalty.

  FitFailedWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_validation.py:536: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
ValueError: Solver lbfgs supports only 'l2' or 'none' penalties, got l1 penalty.

  FitFailedWarning)


[CV]  C=4.281332398719396, penalty=l2, solver=sag, score=0.992, total=   0.0s
[CV] C=4.281332398719396, penalty=l2, solver=sag .....................
[CV]  C=4.281332398719396, penalty=l2, solver=sag, score=0.987, total=   0.0s
[CV] C=4.281332398719396, penalty=l2, solver=lbfgs ...................
[CV]  C=4.281332398719396, penalty=l2, solver=lbfgs, score=0.989, total=   0.0s
[CV] C=4.281332398719396, penalty=l2, solver=lbfgs ...................
[CV]  C=4.281332398719396, penalty=l2, solver=lbfgs, score=0.992, total=   0.0s
[CV] C=4.281332398719396, penalty=l2, solver=lbfgs ...................
[CV]  C=4.281332398719396, penalty=l2, solver=lbfgs, score=0.987, total=   0.0s
[CV] C=11.288378916846883, penalty=l1, solver=liblinear ..............
[CV]  C=11.288378916846883, penalty=l1, solver=liblinear, score=0.993, total=   0.1s
[CV] C=11.288378916846883, penalty=l1, solver=liblinear ..............
[CV]  C=11.288378916846883, penalty=l1, solver=liblinear, score=0.995, total=   0.1s
[CV] C=1

/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_validation.py:536: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
ValueError: Solver sag supports only 'l2' or 'none' penalties, got l1 penalty.

  FitFailedWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_validation.py:536: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
ValueError: Solver lbfgs supports only 'l2' or 'none' penalties, got l1 penalty.

  FitFailedWarning)


[CV]  C=11.288378916846883, penalty=l2, solver=liblinear, score=0.990, total=   0.0s
[CV] C=11.288378916846883, penalty=l2, solver=liblinear ..............
[CV]  C=11.288378916846883, penalty=l2, solver=liblinear, score=0.993, total=   0.0s
[CV] C=11.288378916846883, penalty=l2, solver=liblinear ..............
[CV]  C=11.288378916846883, penalty=l2, solver=liblinear, score=0.989, total=   0.0s
[CV] C=11.288378916846883, penalty=l2, solver=sag ....................
[CV]  C=11.288378916846883, penalty=l2, solver=sag, score=0.990, total=   0.1s
[CV] C=11.288378916846883, penalty=l2, solver=sag ....................
[CV]  C=11.288378916846883, penalty=l2, solver=sag, score=0.993, total=   0.1s
[CV] C=11.288378916846883, penalty=l2, solver=sag ....................
[CV]  C=11.288378916846883, penalty=l2, solver=sag, score=0.990, total=   0.1s
[CV] C=11.288378916846883, penalty=l2, solver=lbfgs ..................
[CV]  C=11.288378916846883, penalty=l2, solver=lbfgs, score=0.990, total=   0.0s
[

/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_validation.py:536: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
ValueError: Solver sag supports only 'l2' or 'none' penalties, got l1 penalty.

  FitFailedWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_validation.py:536: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
ValueError: Solver lbfgs supports only 'l2' or 'none' penalties, got l1 penalty.

  FitFailedWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the

[CV]  C=29.763514416313132, penalty=l2, solver=sag, score=0.994, total=   0.1s
[CV] C=29.763514416313132, penalty=l2, solver=sag ....................
[CV]  C=29.763514416313132, penalty=l2, solver=sag, score=0.991, total=   0.1s
[CV] C=29.763514416313132, penalty=l2, solver=lbfgs ..................
[CV]  C=29.763514416313132, penalty=l2, solver=lbfgs, score=0.990, total=   0.0s
[CV] C=29.763514416313132, penalty=l2, solver=lbfgs ..................
[CV]  C=29.763514416313132, penalty=l2, solver=lbfgs, score=0.994, total=   0.1s
[CV] C=29.763514416313132, penalty=l2, solver=lbfgs ..................
[CV]  C=29.763514416313132, penalty=l2, solver=lbfgs, score=0.991, total=   0.0s
[CV] C=78.47599703514607, penalty=l1, solver=liblinear ...............
[CV]  C=78.47599703514607, penalty=l1, solver=liblinear, score=0.993, total=   0.1s
[CV] C=78.47599703514607, penalty=l1, solver=liblinear ...............
[CV]  C=78.47599703514607, penalty=l1, solver=liblinear, score=0.995, total=   0.1s
[CV] 

/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_validation.py:536: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
ValueError: Solver sag supports only 'l2' or 'none' penalties, got l1 penalty.

  FitFailedWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_validation.py:536: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
ValueError: Solver lbfgs supports only 'l2' or 'none' penalties, got l1 penalty.

  FitFailedWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)


[CV]  C=78.47599703514607, penalty=l2, solver=sag, score=0.990, total=   0.1s
[CV] C=78.47599703514607, penalty=l2, solver=sag .....................
[CV]  C=78.47599703514607, penalty=l2, solver=sag, score=0.994, total=   0.1s
[CV] C=78.47599703514607, penalty=l2, solver=sag .....................


/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)


[CV]  C=78.47599703514607, penalty=l2, solver=sag, score=0.991, total=   0.1s
[CV] C=78.47599703514607, penalty=l2, solver=lbfgs ...................
[CV]  C=78.47599703514607, penalty=l2, solver=lbfgs, score=0.991, total=   0.1s
[CV] C=78.47599703514607, penalty=l2, solver=lbfgs ...................
[CV]  C=78.47599703514607, penalty=l2, solver=lbfgs, score=0.994, total=   0.1s
[CV] C=78.47599703514607, penalty=l2, solver=lbfgs ...................
[CV]  C=78.47599703514607, penalty=l2, solver=lbfgs, score=0.991, total=   0.1s
[CV] C=206.913808111479, penalty=l1, solver=liblinear ................
[CV]  C=206.913808111479, penalty=l1, solver=liblinear, score=0.993, total=   0.1s
[CV] C=206.913808111479, penalty=l1, solver=liblinear ................
[CV]  C=206.913808111479, penalty=l1, solver=liblinear, score=0.995, total=   0.1s
[CV] C=206.913808111479, penalty=l1, solver=liblinear ................
[CV]  C=206.913808111479, penalty=l1, solver=liblinear, score=0.995, total=   0.1s
[CV] C=

/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_validation.py:536: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
ValueError: Solver sag supports only 'l2' or 'none' penalties, got l1 penalty.

  FitFailedWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_validation.py:536: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
ValueError: Solver lbfgs supports only 'l2' or 'none' penalties, got l1 penalty.

  FitFailedWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the

[CV]  C=206.913808111479, penalty=l2, solver=sag, score=0.994, total=   0.1s
[CV] C=206.913808111479, penalty=l2, solver=sag ......................
[CV]  C=206.913808111479, penalty=l2, solver=sag, score=0.992, total=   0.1s
[CV] C=206.913808111479, penalty=l2, solver=lbfgs ....................
[CV]  C=206.913808111479, penalty=l2, solver=lbfgs, score=0.992, total=   0.1s
[CV] C=206.913808111479, penalty=l2, solver=lbfgs ....................
[CV]  C=206.913808111479, penalty=l2, solver=lbfgs, score=0.995, total=   0.0s
[CV] C=206.913808111479, penalty=l2, solver=lbfgs ....................
[CV]  C=206.913808111479, penalty=l2, solver=lbfgs, score=0.994, total=   0.1s
[CV] C=545.5594781168514, penalty=l1, solver=liblinear ...............
[CV]  C=545.5594781168514, penalty=l1, solver=liblinear, score=0.993, total=   0.1s
[CV] C=545.5594781168514, penalty=l1, solver=liblinear ...............
[CV]  C=545.5594781168514, penalty=l1, solver=liblinear, score=0.995, total=   0.1s
[CV] C=545.5594

/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_validation.py:536: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
ValueError: Solver sag supports only 'l2' or 'none' penalties, got l1 penalty.

  FitFailedWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_validation.py:536: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
ValueError: Solver lbfgs supports only 'l2' or 'none' penalties, got l1 penalty.

  FitFailedWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)


[CV]  C=545.5594781168514, penalty=l2, solver=sag, score=0.991, total=   0.1s
[CV] C=545.5594781168514, penalty=l2, solver=sag .....................
[CV]  C=545.5594781168514, penalty=l2, solver=sag, score=0.994, total=   0.1s
[CV] C=545.5594781168514, penalty=l2, solver=sag .....................


/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)


[CV]  C=545.5594781168514, penalty=l2, solver=sag, score=0.992, total=   0.1s
[CV] C=545.5594781168514, penalty=l2, solver=lbfgs ...................
[CV]  C=545.5594781168514, penalty=l2, solver=lbfgs, score=0.992, total=   0.1s
[CV] C=545.5594781168514, penalty=l2, solver=lbfgs ...................
[CV]  C=545.5594781168514, penalty=l2, solver=lbfgs, score=0.995, total=   0.1s
[CV] C=545.5594781168514, penalty=l2, solver=lbfgs ...................
[CV]  C=545.5594781168514, penalty=l2, solver=lbfgs, score=0.995, total=   0.1s
[CV] C=1438.44988828766, penalty=l1, solver=liblinear ................
[CV]  C=1438.44988828766, penalty=l1, solver=liblinear, score=0.993, total=   0.1s
[CV] C=1438.44988828766, penalty=l1, solver=liblinear ................
[CV]  C=1438.44988828766, penalty=l1, solver=liblinear, score=0.995, total=   0.1s
[CV] C=1438.44988828766, penalty=l1, solver=liblinear ................
[CV]  C=1438.44988828766, penalty=l1, solver=liblinear, score=0.995, total=   0.1s
[CV] C=

/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_validation.py:536: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
ValueError: Solver sag supports only 'l2' or 'none' penalties, got l1 penalty.

  FitFailedWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_validation.py:536: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
ValueError: Solver lbfgs supports only 'l2' or 'none' penalties, got l1 penalty.

  FitFailedWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)


[CV]  C=1438.44988828766, penalty=l2, solver=sag, score=0.991, total=   0.1s
[CV] C=1438.44988828766, penalty=l2, solver=sag ......................
[CV]  C=1438.44988828766, penalty=l2, solver=sag, score=0.994, total=   0.1s
[CV] C=1438.44988828766, penalty=l2, solver=sag ......................


/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)


[CV]  C=1438.44988828766, penalty=l2, solver=sag, score=0.992, total=   0.1s
[CV] C=1438.44988828766, penalty=l2, solver=lbfgs ....................
[CV]  C=1438.44988828766, penalty=l2, solver=lbfgs, score=0.993, total=   0.1s
[CV] C=1438.44988828766, penalty=l2, solver=lbfgs ....................
[CV]  C=1438.44988828766, penalty=l2, solver=lbfgs, score=0.995, total=   0.1s
[CV] C=1438.44988828766, penalty=l2, solver=lbfgs ....................
[CV]  C=1438.44988828766, penalty=l2, solver=lbfgs, score=0.995, total=   0.1s
[CV] C=3792.690190732246, penalty=l1, solver=liblinear ...............
[CV]  C=3792.690190732246, penalty=l1, solver=liblinear, score=0.993, total=   0.1s
[CV] C=3792.690190732246, penalty=l1, solver=liblinear ...............
[CV]  C=3792.690190732246, penalty=l1, solver=liblinear, score=0.995, total=   0.1s
[CV] C=3792.690190732246, penalty=l1, solver=liblinear ...............
[CV]  C=3792.690190732246, penalty=l1, solver=liblinear, score=0.995, total=   0.1s
[CV] C=3

/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_validation.py:536: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
ValueError: Solver sag supports only 'l2' or 'none' penalties, got l1 penalty.

  FitFailedWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_validation.py:536: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
ValueError: Solver lbfgs supports only 'l2' or 'none' penalties, got l1 penalty.

  FitFailedWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the

[CV]  C=3792.690190732246, penalty=l2, solver=sag, score=0.994, total=   0.1s
[CV] C=3792.690190732246, penalty=l2, solver=sag .....................
[CV]  C=3792.690190732246, penalty=l2, solver=sag, score=0.992, total=   0.1s
[CV] C=3792.690190732246, penalty=l2, solver=lbfgs ...................
[CV]  C=3792.690190732246, penalty=l2, solver=lbfgs, score=0.993, total=   0.1s
[CV] C=3792.690190732246, penalty=l2, solver=lbfgs ...................
[CV]  C=3792.690190732246, penalty=l2, solver=lbfgs, score=0.995, total=   0.1s
[CV] C=3792.690190732246, penalty=l2, solver=lbfgs ...................
[CV]  C=3792.690190732246, penalty=l2, solver=lbfgs, score=0.995, total=   0.1s
[CV] C=10000.0, penalty=l1, solver=liblinear .........................
[CV]  C=10000.0, penalty=l1, solver=liblinear, score=0.993, total=   0.1s
[CV] C=10000.0, penalty=l1, solver=liblinear .........................
[CV]  C=10000.0, penalty=l1, solver=liblinear, score=0.995, total=   0.1s
[CV] C=10000.0, penalty=l1, so

/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_validation.py:536: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
ValueError: Solver sag supports only 'l2' or 'none' penalties, got l1 penalty.

  FitFailedWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_validation.py:536: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
ValueError: Solver lbfgs supports only 'l2' or 'none' penalties, got l1 penalty.

  FitFailedWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)


[CV]  C=10000.0, penalty=l2, solver=liblinear, score=0.995, total=   0.0s
[CV] C=10000.0, penalty=l2, solver=sag ...............................
[CV] ... C=10000.0, penalty=l2, solver=sag, score=0.991, total=   0.1s
[CV] C=10000.0, penalty=l2, solver=sag ...............................


/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)


[CV] ... C=10000.0, penalty=l2, solver=sag, score=0.994, total=   0.1s
[CV] C=10000.0, penalty=l2, solver=sag ...............................
[CV] ... C=10000.0, penalty=l2, solver=sag, score=0.992, total=   0.1s
[CV] C=10000.0, penalty=l2, solver=lbfgs .............................
[CV] . C=10000.0, penalty=l2, solver=lbfgs, score=0.993, total=   0.1s
[CV] C=10000.0, penalty=l2, solver=lbfgs .............................
[CV] . C=10000.0, penalty=l2, solver=lbfgs, score=0.995, total=   0.1s
[CV] C=10000.0, penalty=l2, solver=lbfgs .............................
[CV] . C=10000.0, penalty=l2, solver=lbfgs, score=0.995, total=   0.1s


[Parallel(n_jobs=1)]: Done 360 out of 360 | elapsed:   10.6s finished


GridSearchCV(cv=3, error_score=nan,
             estimator=LogisticRegression(C=1.0, class_weight=None, dual=False,
                                          fit_intercept=True,
                                          intercept_scaling=1, l1_ratio=None,
                                          max_iter=100, multi_class='auto',
                                          n_jobs=None, penalty='l2',
                                          random_state=42, solver='lbfgs',
                                          tol=0.0001, verbose=0,
                                          warm_start=False),
             iid='deprecated', n_jobs=None,
             param_grid={'C': array([1.00000000e-04, 2.636...
       4.83293024e-03, 1.27427499e-02, 3.35981829e-02, 8.85866790e-02,
       2.33572147e-01, 6.15848211e-01, 1.62377674e+00, 4.28133240e+00,
       1.12883789e+01, 2.97635144e+01, 7.84759970e+01, 2.06913808e+02,
       5.45559478e+02, 1.43844989e+03, 3.79269019e+03, 1.00000000e+04]),
      

In [35]:
grid_search.best_params_

{'C': 11.288378916846883, 'penalty': 'l1', 'solver': 'liblinear'}

In [36]:
grid_search.best_estimator_.fit(X_strain,y_train)

LogisticRegression(C=11.288378916846883, class_weight=None, dual=False,
                   fit_intercept=True, intercept_scaling=1, l1_ratio=None,
                   max_iter=100, multi_class='auto', n_jobs=None, penalty='l1',
                   random_state=42, solver='liblinear', tol=0.0001, verbose=0,
                   warm_start=False)

In [37]:
y_pred = grid_search.best_estimator_.predict(X_stest)

In [38]:
confusion_matrix(y_test,y_pred)

array([[333,   8],
       [  4, 955]])

In [39]:
classification_report(y_test,y_pred)

'              precision    recall  f1-score   support\n\n           0       0.99      0.98      0.98       341\n           1       0.99      1.00      0.99       959\n\n    accuracy                           0.99      1300\n   macro avg       0.99      0.99      0.99      1300\nweighted avg       0.99      0.99      0.99      1300\n'

           precision    recall  f1-score   support

           0       0.99      0.98      0.98       341
           1       0.99      1.00      0.99       959

    accuracy                           0.99      1300
   macro avg       0.99      0.99      0.99      1300

weighted avg       0.99      0.99      0.99      1300 

In [40]:
grid_search.score

<bound method BaseSearchCV.score of GridSearchCV(cv=3, error_score=nan,
             estimator=LogisticRegression(C=1.0, class_weight=None, dual=False,
                                          fit_intercept=True,
                                          intercept_scaling=1, l1_ratio=None,
                                          max_iter=100, multi_class='auto',
                                          n_jobs=None, penalty='l2',
                                          random_state=42, solver='lbfgs',
                                          tol=0.0001, verbose=0,
                                          warm_start=False),
             iid='deprecated', n_jobs=None,
             param_grid={'C': array([1.00000000e-04, 2.636...
       4.83293024e-03, 1.27427499e-02, 3.35981829e-02, 8.85866790e-02,
       2.33572147e-01, 6.15848211e-01, 1.62377674e+00, 4.28133240e+00,
       1.12883789e+01, 2.97635144e+01, 7.84759970e+01, 2.06913808e+02,
       5.45559478e+02, 1.43844989e+03, 3.79

In [41]:
grid_search.cv_results_

{'mean_fit_time': array([0.00624816, 0.00026019, 0.00025137, 0.00553091, 0.01900729,
        0.01049956, 0.00391444, 0.00027084, 0.00026155, 0.00832629,
        0.01925723, 0.00895818, 0.00390124, 0.00026846, 0.00027593,
        0.00570671, 0.01939185, 0.00927051, 0.00472728, 0.00026234,
        0.00028332, 0.00785303, 0.01852512, 0.01001803, 0.01228054,
        0.00027593, 0.00025479, 0.00628026, 0.01592938, 0.01318431,
        0.00750883, 0.00028372, 0.00026147, 0.008413  , 0.01497801,
        0.01656262, 0.0111479 , 0.0003047 , 0.00026854, 0.00744216,
        0.01516747, 0.01981624, 0.02167455, 0.00029985, 0.0002691 ,
        0.00911999, 0.01514268, 0.02250012, 0.04848099, 0.00029683,
        0.00026584, 0.00979479, 0.01815486, 0.03686786, 0.07240327,
        0.00032997, 0.00027124, 0.01056576, 0.01960039, 0.03535501,
        0.07751012, 0.00029794, 0.0002803 , 0.01303172, 0.02407503,
        0.03281434, 0.09007176, 0.00029071, 0.00027124, 0.01066931,
        0.03105275, 0.03966244,

In [42]:
grid_search.best_score_

0.9944201163218899